Added 3-rd level headings for everything

jypma opened the pull request:
This makes beamer export behave a lot better.

jypma committed:
Added 3-rd level headings for everything

This makes beamer export behave a lot better.

jypma commented:
Here's the PDF that beamer creates from this:
[LB3210.pdf](https://github.com/jypma/LB3210/files/6545206/LB3210.pdf)

jypma committed:
Describe Databases

jypma committed:
Describe RabbitMQ

jypma committed:
Describe REST versioning

jypma commented:
[LB3210.pdf](https://github.com/jypma/LB3210/files/6552286/LB3210.pdf)
new version

jbendsen commented on LB3210.org:
```diff
@@ -618,6 +618,92 @@ public record Rectangle(float length, float width) {
+}
+#+END_SRC
+- Spring will send a consumer =ack= when, and if, method returns normally
+- Exceptions will cause a =nack= with redelivery (except for some specific cases)
```
Explain meaning of nack (negative acknowledge). It may not be common knowledge.

jypma replied:
That's explained in the section about RabbitMQ itself, earlier up.

jbendsen commented on LB3210.org:
```diff
@@ -618,6 +618,92 @@ public record Rectangle(float length, float width) {
+</dependency>
+#+END_SRC
+
+- Will look for rabbitMQ locally on the default port
```
What will happen if rabbitmq is down when app starts? What is best practice here? I guess the same question is relevant for databases.

jypma replied:
It will start normally (that's stated later). And that's not best practice, but I wasn't able to coerce Spring Boot to do otherwise.

jbendsen commented on LB3210.org:
```diff
@@ -788,10 +788,46 @@ Route route = path("users", () ->
+
+- Manage your migrations: =Flyway=
+  * Write incremental changes as scripts
+  * Trace script execution in the database itself
```
You could consider adding that Flyway only support forward changes as opposed to other tools, e.g. LuiquidDB. The idea is to back up the database prior to flyway script execution and restore it if rollback is needed.

jypma replied:
Hmm let's have a talk about that. I explicitly recommend forward-only changes as the "recommended" way

jbendsen commented on LB3210.org:
```diff
@@ -618,6 +618,92 @@ public record Rectangle(float length, float width) {
+- Type-safe table and column definitions
+- Automatic mapping into generated class
+- Compile errors when you remove/change columns in your schema
+** RabbitMQ
```
Do you have some general introduction to RabbitMQ? What is the main reason to use Rabbit?

jypma replied:
The general introduction to Rabbit is earlier in our presentation.

The reason to mention Rabbit is because this customer requested it, but more reasoning is in the earlier RabbitMQ heading.

jbendsen commented on LB3210.org:
```diff
@@ -977,6 +982,208 @@ Route route = path("users", () ->
+* A selection of REST patterns
+** Resource tags and caching
+*** Resources have versions
+- Servers can include an =ETag=, which specifies which /version/ of a resource is being served
```
Consider explaining: ETag = Entity Tag, Entity~The 'thing' a Resource represents.  As fielding states it: 'More precisely, a resource R is a temporally varying membership function MR(t), which for time t maps to a set of entities, or values, which are equivalent'.

jbendsen commented on LB3210.org:
```diff
@@ -977,6 +982,208 @@ Route route = path("users", () ->
+  * =application/json=
+  * =application/protobuf=
+- But it doesn't have to be
+  * =application/vnd.example.myresource.v1+json=
```
Consider: vnd=vendor. A vendor specific format as opposed to standardized formats.

jypma committed:
Add section on Async Java

This is added in the project directory, so emacs jshell integration can
pick it up when live coding.

jbendsen commented on LB3210.org:
```diff
@@ -977,6 +982,208 @@ Route route = path("users", () ->
+- Server sends a custom header of the API version that's current
+- This does kinda work
+
+- Fairly weak way to work around /actually/ dealing with semantic changes and compatibility
```
consider adding references and examples to best-of-breed API's our there, e.g. Github, tradeshift (I guess you know this one pretty well:-), e-conomic, stripe.com.

jypma committed:
Final corrections, add some fun graphics
